### ISSUE TITLE
Read Timeout Exception leaves connection in a bad state

### ISSUE THREAD

**Issue by nicofff** [phpredis/phpredis#1222]
If a read timeout exception is triggered, and caught, the connection remains in a weird state were subsequent requests will return bad data.

This issue seems to be exclusive to RedisCluster. I haven't been able to trigger it connecting to a normal Redis server.

- OS: ArchLinux, CentOS
- Redis: 3.0.6, 3.2.5, 4.0.1
- PHP: 7.0.21, 7.1.8 
- phpredis: 3.1.2, 3.1.3, develop

Tested with Cluster Nodes on local machine, in a vm and in different instances in AWS.

Steps to reproduce:

See attached files. [read-timeout-repro.zip](https://github.com/phpredis/phpredis/files/1229369/read-timeout-repro.zip)

The ClusterTest.php sets some keys with know values and then reads them randomly and checks that the value read is the expected one
The fail.sh script uses iptables to force a read timeout

Steps (assuming local cluster):
1) Edit ClusterTest.php to point to the cluster nodes
2) Edit fail.sh, set port to where a master node is running
3) Run ClusterTest.php from a shell
4) Wait until it outputs "Testing!"
5) Run fail.sh 
6) The ClusterTest.php script will start throwing errors about reading values that didn't match what was expected

The issue persist until the connection is closed. 
If persistent connections are not used, closing and reopening the connection will return a connection in a good state.
If persistent connections are used, the connection remains in a bad state even after calling close()

**Comment by michael-grunder** [comment]
Hey thanks for the report.  I'll take a look this evening and see if I can replicate the problem but having written the cluster code think you are likely correct.  Getting that right on a distributed system was a bit tricky.  :smile:

**Comment by sebastiank** [comment]
Hi Michael, did you have time to look that issue ?

**Comment by michael-grunder** [comment]
Hey,

I took a look when you first posted the issue and was actually able to reproduce the problem in a few different ways.  One way for example is by just issuing a `DEBUG SLEEP` to one of the cluster instances.

I've been pretty busy at work but am going to try to spend some time looking for a reasonable solution to the problem over the weekend.

**Comment by HDANILO** [comment]
Do we have update on this? It is only happening to persistent connection on RedisCluster?

**Comment by ErwinSteffens** [comment]
We have had the same issue. Also using persistent connections on RedisCluster. 

We get a `RedisException` with message `socket error on read socket` message. After this we are getting invalid data.

**Comment by HDANILO** [comment]
Did you find why the connection is reading bad data? Perhaps thats what should be fixed, might be a bad offset, not sure.

If you find better go to the close route, perhaps it would be nice to everytime a connection read exception happens, the redis cluster handle it in the background, trying to reconnect 3 times with exponential backoff to avoid throttling, and then if not succeeded(and probably won't because this might be happening because of heavy traffic) close the persistent connection and throw an exception letting the user know that the connection was closed so he can fail gracefully.

**Comment by ErwinSteffens** [comment]
@michael-grunder Did you have any luck in finding this issue? We disabled persistent connections but are getting issues with this.

**Comment by michael-grunder** [comment]
Hey everyone,

I pushed a small commit that should help with this issue.  I'll need to hook this logic into more places in the code but what it does is in the event of a timeout phpredis will disconnect from the node even if using persistent connections.

Then, since the `RedisCluster` class uses lazy connections it will reconnect to the node the next time it needs to.  This should stop phpredis from getting into an inconsistent state -- at least in the case of a timeout.

The next step will be adding this logic to the other places that could trigger this state.

**Comment by HDANILO** [comment]
It still throws the exception right? Otherwise it could be chaotic

**Comment by michael-grunder** [comment]
Yes, it will still throw an exception but also close the connection to that specific node.  Then, the next time data is requested from that node it should reconnect.  The idea here is to avoid phpredis getting out of sync with redis-server.

**Comment by ErwinSteffens** [comment]
Isn't it better to check the outcome of checking the response:
```
resp = cluster_check_response(c, &c->reply_type TSRMLS_CC);
```

This can result in -1 which means the connection is in bad state. The loop still continues until time-out and tries to send commands to all the nodes in the cluster.

**Comment by michael-grunder** [comment]
@ErwinSteffens I got your comment on the change but now can't find it on GitHub!

> Isn't it better to check the outcome of checking the response:
> resp = cluster_check_response(c, &c->reply_type TSRMLS_CC);

Likely yes.  I probably need to implement some sort of connection reset logic for _any_ early return or exception.  This fix is just a start but does seem to work as a proof of concept for handling timeout with persistent connections.  Before the change, I got bad data every time a timeout occurred but not after the update.

**Comment by ErwinSteffens** [comment]
I removed the comment because we first wanted to do some more investigation for ourself to understand the problem.

We will test some of the new code tomorrow. When ok we will put it live so we can re-enable persistent connections.

**Comment by michael-grunder** [comment]
Sounds good.  I will try to push a more robust fix this evening.  Let me know how your testing goes.

**Comment by michael-grunder** [comment]
Hey everyone,

I think I've come up with a solution to this problem that should be pretty robust.

I'm going to add a `dirty` flag to the `redisClusterNode` structure.  We'll mark the node dirty before sending a request and then clean once we successfully get a reply.  The flag will also be reset on a true disconnect/reconnect.

Once that flag is in place, we can just check the flag before attempting communication in our command loop.  If we detect a node dirty, we'll do an actual disconnection before communicating with the node.  This should cover every command including ones that don't go through the normal command loop (e.g. directed commands).

Let me know if this seems like a reasonable solution to everyone.

Cheers!
Mike

**Comment by HDANILO** [comment]
Don't know enough so I'll just ask, its possible to issue a command in the node before the first command gets back? This could cause some issues. For me, it's totally ok as long as this is clear in the documentation.

**Comment by yatsukhnenko** [comment]
@michael-grunder may be this flag useful for Redis and RedisArray objects?

**Comment by michael-grunder** [comment]
@yatsukhnenko That's actually a really good point since we use persistent connections in Redis and RedisArray classes as well.  Perhaps better to add it to `redis_sock` itself.

**Comment by michael-grunder** [comment]
> Don't know enough so I'll just ask, its possible to issue a command in the node before the first command gets back? This could cause some issues. For me, it's totally ok as long as this is clear in the documentation.

@HDANILO  RedisCluster doesn't allow pipelining so every command/reply is handled in sequence.  There are some abstractions (e.g. around MULTI..EXEC) but this strategy should work there too (if we deliver MULTI but don't get the OK, the node will remain dirty).

**Comment by marcdejonge** [comment]
I think adding a `dirty` flag to the `redis_sock` seems a good solution. Maybe you can set this flag only when you expect a response, which is most of the time. At least I would hope you can add it to the RedisCluster as soon as possible, such that we can test it. We really need a solution to be able to use the persistent connections again.

**Comment by marcdejonge** [comment]
I've found something that is causing the connection errors: we were out of ephemeral connections! It seems that phpredis is first opening a connection to get the cluster status and then it closes this connection again immediately. After that, there is a decent change (1 in 3 in our case) that we need a connection to the same server again. So, this seems to me as not very efficient.

Furthermore, these connection errors result in timeouts, while it is simply that it keeps trying to create a connection while it fails. It seems to me that on `library.c:1469` the error number that is returned should be checked. In our case this returns the number 99, which I decoded using http://php.net/manual/en/function.socket-last-error.php to find out what really the error is.

**Comment by michael-grunder** [comment]
I can update `cluster_map_slots` to take the connected seed node as an argument and then attach it to the appropriate node.  That's not a bad idea as it will be a bit more efficient, but only slightly.  I think that updating [this](https://github.com/phpredis/phpredis/tree/cache-slots) branch will provide a more substantial improvement since the slots get cached between requests allowing phpredis to avoid this step altogether.

I'm presently going through all of the cluster code and instrumenting any early returns such that they don't leave the cluster in a bad state.  In addition, I've added an optional boolean argument to `RedisCluster->close()` which allows PHP code to _really_ close cluster connections, even if using persistent connections.

**Comment by marcdejonge** [comment]
Any updates? I'd love to help, also with testing, but I'd have to know which branch to look at and what to do...

**Comment by ErwinSteffens** [comment]
@michael-grunder Have you got any updates on this issue? We are still running with persistent connections turned off in production but it is causing us issues. As @marcdejonge mentioned we would love to help out, but we have to know the current status.

**Comment by joshholat** [comment]
+1, also causing us issues. Seems to happen both with and without persistent connections in our testing. We are on v3.1.6 currently and PHP 7.1. Would be happy to help test any potential updates when they are available.

**Comment by nicofff** [comment]
Hey, I was doing some testing with the latest release and the cluster-persistent-state-flush branch. 
Both of them survive killing the connections with iptables (ie: the script originally attached), but I switched to netem and:
a) On the 4.0.0 release, I can still reproduce the issue (the wrong key is returned)
b) On the cluster-persistent-state-flush branch, after a few timeout exceptions the client errors with a "Can't communicate with any node in the cluster" exception that persists event after I delete the netem rules.

The netem command I used was:
tc qdisc add dev {your_network_interface} root netem delay 100 10

It seems related to package reordering as running the same test without the last param in the netem command (the variable latency, which causes reordering) works fine.

**Comment by ErwinSteffens** [comment]
Hi all, any updates on this issue? We are still running with persistent connection disabled and this is causing some issues at the moment.

We are willing to help with providing information or testing.

**Comment by marcdejonge** [comment]
Still no updates? We are thinking of switching to a different library because of this, it is still causing problems.

**Comment by ErwinSteffens** [comment]
Hi all, we are still having issues with this. Maybe it is unusual to do this, but we are willing to pay for having this issue fixed. If anyone is interested, please let me know.

**Comment by yatsukhnenko** [comment]
> After this we are getting invalid data.
@ErwinSteffens if I understand correctly you have a bit different issue than was described by @nicofff.
Could you explain this?

I'm going to change reconnect logic and change behavior of `close` method (force close connection even if it is persistent). It may be useful to create workaround in your applications while we didn't get it in extension

**Comment by ErwinSteffens** [comment]
No, we have the same issue as described by @nicofff. There is a read time-out with shows by a `RedisException` with message `socket error on read socket` message. 

After this we receive invalid data. It is the same as described in the reproduction in the start of the topic.

**Comment by yatsukhnenko** [comment]
@nicofff @ErwinSteffens I made changes in reconnect logic. Could you test [changes](https://github.com/yatsukhnenko/phpredis)

**Comment by nicofff** [comment]
@yatsukhnenko Tested the develop branch from your fork, still fails as reported originally.

**Comment by yatsukhnenko** [comment]
@nicofff, @ErwinSteffens  I added more changes to close/reconnect logic and push them to [develop](https://github.com/phpredis/phpredis) branch. Now you can use `Redis::close` and `RedisCluster::close` to close persistant connections.

**Comment by nicofff** [comment]
@yatsukhnenko  Rerun the test with the new code you pushed:
1) Without modifying the test (ie: don't call close on exception) I'm consistently getting only one wrongly read key, and then it returns to normal operation, instead of continuing to read wrong values indefinitely.
2) Modifying the test to call close after an exception is triggered does seem to prevent the issue.

**Comment by ErwinSteffens** [comment]
Thanks. We will start testing it next week.

**Comment by yatsukhnenko** [comment]
@ErwinSteffens how is your tests going?

**Comment by yatsukhnenko** [comment]
@ebuildy could you open new issue ?

**Comment by yatsukhnenko** [comment]
@nicofff the issue is resolved? can we close it?

**Comment by ErwinSteffens** [comment]
Sorry, but we had some change in priorities and did not do any testing yet. Hope we can do it next week.

**Comment by nicofff** [comment]
@yatsukhnenko We don't currently have a live deployment with redis cluster. I'd love to wait to see what results @ErwinSteffens gets.

From the testing I made, I'd say the issue impact has been greatly reduced and can be worked around. I wouldn't say it's fixed. I think to consider it fixed, one of two things should happen.
a) The single instance of key misreading after an exception shouldn't happen, if the connection is not closed by the user
b) The connection should be automatically closed after an exception

**Comment by nicofff** [comment]
@ErwinSteffens how did the testing go?

**Comment by marcdejonge** [comment]
Erwin hasn't had time to do it, because it's getting too busy. Right now I'm going to take over. Which version can I use for testing? As far as I can see the version 4.1.1 contains the improved connection handling, right?

**Comment by yatsukhnenko** [comment]
@marcdejonge you should test `develop` branch or wait upcoming `4.2.0` release

**Comment by marcdejonge** [comment]
Hmm, ok... I'll take a look if I can build it somewhere. We're currently still running on version 3.1.6, are there any breaking changes to this?

**Comment by yatsukhnenko** [comment]
> are there any breaking changes to this?

A lot :smile: We have increased major version number because of [that](https://pecl.php.net/package-changelog.php?package=redis&release=4.0.0) changes.

**Comment by marcdejonge** [comment]
Ok, the build has succeeded and fortunately we didn't have any trouble with the upgrade. After some testing I've put it on one of our front-end servers to test in a live environment. I've not seen any problems until now, but I'll keep monitoring today to see if anything happens. I'll let you know!

**Comment by marcdejonge** [comment]
So we've done a bit of a stress-test with 1 server using the new version and 1 server using the old version. During this stress-test the server with the old version had a lot of troubles with lost connections, but the new version with persistent connections had no trouble at all! I'm really happy with this result.

I'm currently doing the implementation in PHP to handle any RedisClusterException to close the persistent connection, so we can finish this off.

When can we expect a real release of this version? Then I don't have to run a custom build version anymore...

**Comment by yatsukhnenko** [comment]
> When can we expect a real release of this version?

I think it will be right after upcoming redis 5.0 because we have new streams api merged into `develop`.

**Comment by yatsukhnenko** [comment]
And we have a practice releasing two or three release candidates before stable release so it will be not earlier than in a month.

**Comment by marcdejonge** [comment]
Even a release of a RC should be fine for us. Right now I've got a version that I've build myself without CI or anything, which doesn't feel good on a production server ;)

**Comment by nicofff** [comment]
Awesome!
@marcdejonge Thanks for testing!
@yatsukhnenko Sorry to insist, but having a data corruption bug with a non documented workaround (other than this issue) isn't great.
Is disconnecting after the exception something that can be done in the library so that there is no need to do it from the user side? If so, do you want me to create another issue for it and close this?
If it's not posible, should I create an issue/ PR to update the documentation?

**Comment by yatsukhnenko** [comment]
> Is disconnecting after the exception something that can be done in the library

@nicofff I think it can be done on the phpredis side but I don't know when I will make this changes (this issue doesn't affect the project on my main work :smile:)
I like an idea to close this and create new issue with new actual description of problem. A note in the documentation also will not be superfluous so could you create both issue and PR?

**Comment by liamsorsby** [comment]
@yatsukhnenko @nicofff @marcdejonge  I've just rebuild the new redis.so library from the develop branch and run the test script above and I can still replicate the issue. However, instead of running iptables we have been incrementally calling the cluster failover on the masters.

**Comment by marcdejonge** [comment]
Well, if you block all traffic it will timeout. But at least it will restore the connection correctly afterwards, this was the real bug. You can't really expect something else than an exception if it can't connect to the master server, right?

**Comment by liamsorsby** [comment]
@marcdejonge I wasn't saying that I got an error/timeout. I was saying that I still get incorrect data returned.

**Comment by liamsorsby** [comment]
![screen shot 2018-10-09 at 12 17 48](https://user-images.githubusercontent.com/5270149/46674401-ec6e7500-cbd3-11e8-9644-440663e9efc7.png)

**Comment by liamsorsby** [comment]
@marcdejonge Yes, I've created a container to pull down the git repo and build from develop. 

`
~/dev/reproduce $ docker-compose up -d php
Creating network "reproduce-bug_default" with driver "bridge"
Creating reproduce-bug_php_1 ... done
~/dev/reproduce $ docker-compose exec php /bin/bash
root@0af7ca5150c4:/var/www/html# php --ri redis

redis

Redis Support => enabled
Redis Version => develop
Git revision => $Id: cd6ebc6d7f56ef652cb1dbe5e02126e95aa30156 $
Available serializers => php

Directive => Local Value => Master Value
redis.arrays.autorehash => 0 => 0
redis.arrays.connecttimeout => 0 => 0
redis.arrays.distributor => no value => no value
redis.arrays.functions => no value => no value
redis.arrays.hosts => no value => no value
redis.arrays.index => 0 => 0
redis.arrays.lazyconnect => 0 => 0
redis.arrays.names => no value => no value
redis.arrays.pconnect => 0 => 0
redis.arrays.previous => no value => no value
redis.arrays.readtimeout => 0 => 0
redis.arrays.retryinterval => 0 => 0
redis.clusters.persistent => 0 => 0
redis.clusters.read_timeout => 0 => 0
redis.clusters.seeds => no value => no value
redis.clusters.timeout => 0 => 0
redis.session.locking_enabled => 0 => 0
redis.session.lock_expire => 0 => 0
redis.session.lock_retries => 10 => 10
redis.session.lock_wait_time => 2000 => 2000
`

**Comment by liamsorsby** [comment]
Run over a period ~2 minutes:

```
Testing!
Wrong value found!!
Key: 118
Value: 26
Wrong value found!!
Key: 37
Value: 55
Wrong value found!!
Key: 109
Value: 126
Wrong value found!!
Key: 67
Value: 71
Wrong value found!!
Key: 85
Value: 27
^C
root@0af7ca5150c4:/var/www/html# php --ri redis | grep Version
Redis Version => develop
root@0af7ca5150c4:/var/www/html# php --ri redis | grep "Git revision"
Git revision => $Id: cd6ebc6d7f56ef652cb1dbe5e02126e95aa30156 $
root@0af7ca5150c4:/var/www/html# php -v
PHP 7.2.10 (cli) (built: Sep 15 2018 02:28:03) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
root@0af7ca5150c4:/var/www/html#
```

**Comment by marcdejonge** [comment]
I'm not able to reproduce it, while I'm using the same PHP version and the same release of phpredis. Can you share your docker image? And the script you are using to trigger this error?

**Comment by liamsorsby** [comment]
Sure. @marcdejonge 
https://github.com/liamsorsby/bug-reproduction
Run docker-compose up -d
in a new terminal run docker-compose exec php /bin/bash
in another terminal run docker-compose run breaker
Then in the php container shell run php Cluster.php

**Comment by marcdejonge** [comment]
Hmm... have been running the script for about 5 minutes, no errors yet. @yatsukhnenko are you able to reproduce this error? Maybe my Virtual Machine is just too slow or something...

**Comment by liamsorsby** [comment]
Interesting. Are you running the failover in the background as well?

**Comment by michael-grunder** [comment]
I can't replicate the problem in that container either.  It doesn't mean that there isn't an issue, but it probably means that it's going to be difficult to find.

**Comment by liamsorsby** [comment]
I'll try in an hour. It may be cached in the apache module. I'll kill it and rebuild it from scratch. The output above is from the cli

**Comment by liamsorsby** [comment]
@michael-grunder @marcdejonge I've used a timer to guage how many transactions per second my laptop is able to process just incase it is based on speed. (The 3 high numbers you can see is before the failovers kick in)
```
Testing!
Number of iterations over 1 seconds: 6577
Number of iterations over 1 seconds: 3370
Number of iterations over 1 seconds: 4
Number of iterations over 1 seconds: 18
Number of iterations over 1 seconds: 2
Number of iterations over 1 seconds: 11
Number of iterations over 1 seconds: 1
Wrong value found!!
Key: 27
Value: 23
Number of iterations over 1 seconds: 19
Number of iterations over 1 seconds: 6
Number of iterations over 1 seconds: 39
Number of iterations over 1 seconds: 1
Number of iterations over 1 seconds: 21
Number of iterations over 1 seconds: 2
Wrong value found!!
Key: 22
Value: 108
Number of iterations over 1 seconds: 21
Number of iterations over 1 seconds: 5
Number of iterations over 1 seconds: 3
Number of iterations over 1 seconds: 4
Number of iterations over 1 seconds: 10
Number of iterations over 1 seconds: 8
Number of iterations over 1 seconds: 26
Number of iterations over 1 seconds: 2
Number of iterations over 1 seconds: 3939
Number of iterations over 1 seconds: 958
Number of iterations over 1 seconds: 11
Number of iterations over 1 seconds: 28
Number of iterations over 1 seconds: 3
Wrong value found!!
Key: 53
Value: 87
Number of iterations over 1 seconds: 24
Number of iterations over 1 seconds: 10
```

**Comment by liamsorsby** [comment]
Is there any sort of profiling I would be able to do to help debug the issue?

**Comment by marcdejonge** [comment]
Can you try this branch? https://github.com/marcdejonge/phpredis/tree/extra-logging

I've added some trace logging to see connection behavior

**Comment by liamsorsby** [comment]
Having a go now @marcdejonge

**Comment by liamsorsby** [comment]
I still replicated the bug @marcdejonge but never saw anything to indicate it had failed over:
```
Number of iterations over 1 seconds: 10
cluster_sock_write - No Failover - Success
cluster_send_command - LOOP 0 / 0 / 0
cluster_sock_write - No Failover - Success
cluster_send_command - LOOP 0 / 0 / 0
Wrong value found!!
Key: 85
Value: 4
cluster_sock_write - No Failover - Success
```

**Comment by liamsorsby** [comment]
@michael-grunder Firstly, I'm glad you can reproduce it as I thought there was an issue with the container. However, regardless of the reproduction steps if there is potential for data to be returned incorrectly it's still a major issue.

If you can point me to a good point to help with the debugging. I'm happy to do so.

**Comment by yatsukhnenko** [comment]
> it's still a major issue

When you need special environment to reproduce issue and it appears one time in ten thousands it doesn't look important (kidding) :smile: 
I'll install docker on my machine and will try to understand what's happening.

**Comment by yatsukhnenko** [comment]
About one hour of testing and there are no errors...

**Comment by yatsukhnenko** [comment]
@liamsorsby do you have this issue in production?

**Comment by liamsorsby** [comment]
@yatsukhnenko It didn't make it to production as we load tested this pre production and could reproduce it. Unfortunately, We now run into to exhaustion as a result of disabling persistence.

P.s. we don't use Docker that was just to reproduce the issue locally.

**Comment by michael-grunder** [comment]
@liamsorsby I've opened a new issue for this problem in #1425.  It's related to this bug but not exactly the same.
